nixos/modules/services/web-apps/librespeed.nix

```diff
@@ -0,0 +1,452 @@
+            "/backend/" = {
+              proxyPass = "http://127.0.0.1:${toString cfg.settings.listen_port}/backend/";
+              extraConfig = # nginx
+                ''
+                  brotli off;
+                  client_max_body_size 0;
+                  gzip off;
+                  proxy_request_buffering off;
+                '';
+            };
+          };
+        };
+      };
+    };
+
+    systemd.services = {
+      librespeed-secrets = lib.mkIf (cfg.secrets != { }) {
+        description = "LibreSpeed secret helper";
+
+        ExecStart = lib.getExe (
+          pkgs.writeShellApplication {
+            name = "librespeed-secrets";
+            runtimeInputs = [ pkgs.coreutils ];
+            text = ''
+              cp ${configFile} ''${RUNTIME_DIRECTORY%%:*}/config.toml
+            ''
+            + lib.pipe cfg.secrets [
+              (lib.mapAttrs (
+                name: file: ''
+                  cat >>''${RUNTIME_DIRECTORY%%:*}/config.toml <<EOF
+                  ${name}="$(<${lib.escapeShellArg file})"
+                  EOF
+                ''
+              ))
+              (lib.concatLines lib.attrValues)
+            ];
+          }
+        );
+
+        serviceConfig = {
+          Type = "oneshot";
+          RemainAfterExit = true;
+          RuntimeDirectory = "librespeed";
+          UMask = "u=rw";
+        };
+      };
+
+      librespeed = {
+        description = "LibreSpeed server daemon";
+        wantedBy = [ "multi-user.target" ];
+        wants = [
+          "network.target"
+        ]
+        ++ lib.optionals (cfg.useACMEHost != null) [
+          "acme-finished-${cfg.useACMEHost}.target"
+        ];
+        requires = lib.optionals (cfg.secrets != { }) [ "librespeed-secrets.service" ];
+        after = [
+          "network.target"
+        ]
+        ++ lib.optionals (cfg.secrets != { }) [
+          "librespeed-secrets.service"
+        ]
+        ++ lib.optionals (cfg.useACMEHost != null) [
+          "acme-finished-${cfg.useACMEHost}.target"
+        ];
+
+        serviceConfig = {
+          Type = "simple";
+          Restart = "always";
+
+          DynamicUser = true;
+
+          LoadCredential = lib.mkIf (cfg.tlsCertificate != null && cfg.tlsKey != null) [
+            "cert.pem:${cfg.tlsCertificate}"
+            "key.pem:${cfg.tlsKey}"
+          ];
+
+          ExecStartPre = lib.mkIf cfg.downloadIPDB "${lib.getExe cfg.package} --update-ipdb";
+          ExecStart = "${lib.getExe cfg.package} -c ${
+            if (cfg.secrets == { }) then configFile else "\${RUNTIME_DIRECTORY%%:*}/config.toml"
+          }";
+          WorkingDirectory = "/var/cache/librespeed";
+          RuntimeDirectory = "librespeed";
+          RuntimeDirectoryPreserve = true;
+          StateDirectory = "librespeed";
+          CacheDirectory = "librespeed";
+          SyslogIdentifier = "librespeed";
+
+          ReadOnlyPaths = [ cfg.package ];
```

Comment:
I am a bit confused by this line: is there some chroot stuff going on so that the package wouldn't already be visible to the service? It's in the nix store, so it should be readonly, right?